test/test_type_promotion.py

```diff
@@ -430,6 +431,97 @@ def test_transpose(self, device):
         a = torch.tensor([[True, True], [False, True]], device=device)
         self.assertEqual(a.t() == 0, a.t() == False)  # noqa: E712
 
+    def get_sparse_tensors(self, device, dtype, coalesced, zeros=True):
+        t = self._get_test_tensor(device, dtype, not zeros)
+        if coalesced:
+            s = t.to_sparse()
+        else:
+            s = t.to_sparse()
+            indices = torch.cat((s.indices(), s.indices()), 1)
+            values = torch.cat((s.values(), s.values()), 0)
+            s = torch.sparse_coo_tensor(indices=indices, values=values, size=s.size(), dtype=dtype, device=device)
+            t = s.to_dense()
+        return (t, s)
+
+    def _get_precision(self, dtype, coalesced):
+        if dtype == torch.half and not coalesced:
+            # very low precision for uncoalesced float16 sparse tensors since
+            # ops like (s1 + s2).to_dense() will add four low-precision
+            # floating point values.
+            return 5e-2
+        if dtype == torch.half:
+            return 1e-3
+        return 1e-5
+
+    def _test_sparse_op(self, op_name, inplace, dtype, device, coalesced):
+        suffix = '_' if inplace else ''
+        err = "{} {}({}, {})".format("  coalesced" if coalesced else "uncoalesced", op_name + suffix, dtype, dtype)
+
+        def op(t1, t2):
+            return getattr(t1, op_name + suffix)(t2)
+
+        add_sub = op_name == 'add' or op_name == 'sub'
+
+        (dense1, sparse1) = self.get_sparse_tensors(device, dtype, coalesced)
+        (dense2, sparse2) = self.get_sparse_tensors(device, dtype, coalesced, op_name != 'div')
+        common_dtype = torch.result_type(dense1, dense2)
+        if self.device_type == 'cpu' and common_dtype == torch.half:
+            self.assertRaises(RuntimeError, lambda: op(s1, d2))
+
+        expected = op(dense1.clone(), dense2)
+        precision = self._get_precision(expected.dtype, coalesced)
+        test_tensors = [expected, dense1, sparse1, dense2, sparse2]
+        e, d1, s1, d2, s2 = [x.clone() for x in test_tensors] if inplace else test_tensors
+
+        # Test op(sparse, sparse)
+        if op_name != 'div':
+            sparse = op(s1, s2)
+            self.assertEqual(sparse.dtype, e.dtype)
+            self.assertEqual(e, sparse.to_dense(), prec=precision, message=err)
+        else:
+            # sparse division only supports division by a scalar
+            self.assertRaises(RuntimeError, lambda: op(s1, s2).to_dense())
+
+        # Test op(dense, sparse)
+        if add_sub:
```

Comment:
There's a lot of special casing going on in this function to handle testing for 4 operators: add, sub, mul, div. The special casing makes it so that it is hard to tell whether or not the code is comprehensively testing each operator. It might be better to break it up into 4 tests "test_add",  "test_sub", "test_mul", "test_div".

Furthermore, these tests should exist in `test_sparse` because they don't do any type promotion

Reply:
I think this actually makes it easier to see that every case is captured, since every `else` is an `assertRaises`. We're negatively affirming all of the uncaptured conditions, which ensures that every supported possibility is tested. If I use distinct test_mul/test_div functions, it's easy to forget one negative assertion for something like a mul(sparse,sparse) case. I'd like to be able to enforce that as we improve sparse support, we are forced to fix the tests.

I'm also afraid that if I break them apart into separate functions I'll either be re-constructing cuda tensors for each operator (which could be slow) or passing around too many parameters for it to be clearly readable.

To preserve the negative tests, if you really think it would be more readable even with many parameters being passed to each test, I'd probably prefer to break it up by dense_sparse, sparse_sparse, etc, instead of by operator.

In other words, this section would look something like:
```
test_sparse_sparse(op_name, inplace, s1, s2, e)
test_dense_sparse(op_name, inplace, d1, s2, e)
test_sparse_scalar(....)
```
the implementations would probably end up repeating some code like the `suffix = '_' if inplace else ''` above.

Or I could just make them nested functions that close over the variables in the setup part, if you think that would be clearer?

Reply:
Here's one example of the things that seem weird to me. We call:
```
dtypes = torch.testing.get_all_math_dtypes(device)
        ops = ['add', 'sub', 'mul', 'div']
        for dtype in dtypes:
            for op_name in ops:
                for inplace, coalesced in itertools.product([True, False], [True, False]):
                    self._test_sparse_op(op_name, inplace, dtype, device, coalesced)
```
Okay, this seems reasonable. I would expect that we are testing the same thing for each operator. 

It turns out, that expectation isn't true. There's a special case for division where we are only testing division by a scalar but not division of two sparse tensors. If someone comes along in the future and adds support for division of two sparse tensors, where do they write a new test case? They would have to:
1) know about the existence of this test (it isn't clear from the test name that this is testing sparse division)
2) read through this function to find where to enable the test for sparse/sparse division.

If, instead, we had a `test_div_sparse_sparse(self)` test case, then a future developer could just add their test case in there and not have to worry about understanding how `test_sparse_ops` is doing all of its checks.

Reply:
> There's a special case for division where we are only testing division by a scalar but not division of two sparse tensors.

We are though, we're explicitly testing the current behavior of division of two sparse tensors, where the current behavior is to raise an error. If someone changes that behavior (by implementing sparse division), they'll be forced to change the test (since it'll start failing when it no longer raises an error). That's a testing style @gchanan often requests for unsupported operations.

Reply:
I feel like we're discussing a few points at once, so let's just go through them one by one.

First, could we move this test case to test_sparse.py? It doesn't relate to type promotion and all the other tests for binary functions like this exist in test_sparse.

Reply:
> First, could we move this test case to test_sparse.py

It will be testing mixed-types though in the next diff. When I initially wrote this test, I actually skipped same-type combinations, and only re-added them for this diff when I came across this bug. Is it really worth moving now just to move back in the next diff, or would you want it to stay in test_sparse even when it is updated to test mixed-types?

Reply:
Since this diff is fixing uncoalesced division, it makes more sense to me to add a single test case for uncoalesced division into test_sparse, and move all of the type promotion testing code into the next diff.

Reply:
I tried to write up what I think would be a clearer alternative to the test code here, but it doesn't save any LOC and also has its tradeoffs :/ https://gist.github.com/zou3519/3c6248a7e2267c796043b6ef59561981 . 

I don't know if it is better, but this is what I mean by wanting something clearer that doesn't have special casing:
- there is a clear `test` function just does what it is supposed to (get test inputs, run the test, exit)
- the table has what is supported, what isn't supported, and which test cases need inputs without zeros.

Reply:
> Since this diff is fixing uncoalesced division

I'd generally favor readability of code as-is over readability of the diff. I think if we can create a set of test that completely covers all cases, it's unhelpful to retain the narrow individual cases that overlap with the broader tests. Statically, the narrow tests will only really be useful to someone looking back at old diffs/issues, and doesn't say much about the current state of the code as it exists fully merged without the bugs present.

I'd actually like to get rid of most of the tests in test_type_promotion that overlap with test_many_promotions, at least to the extent that those tests are testing operations that are fully tested by comprehensive tests. Some older tests like `test_int_promotion` that only existed due to particular examples in historical discussions aren't really helpful to understanding the current state of the code as it exists now. If I landed the narrow test just to check the one specific bug I discovered with these tests, I'd kind of prefer to remove that no-longer-needed test once we had complete coverage in the next diff anyway.

> I tried to write up ...

I find this a bit confusing, so I think this might just be a stylistic difference. My main concern with this approach though would be having the list of tests hard-coded, since it leaves a lot of room for missing a case. There's a fair bit that would also still need to be added too (e.g. can_cast errors in the next diff are dependent on the dtypes of the operands), so I think this would end up being just as much code. Plus by being specific the comments can describe each particular case in the right place, and by re-using test tensors created for out of place operations, we save a bit of run-time over creating new test tensors for each test.

Reply:
> If I landed the narrow test just to check the one specific bug I discovered with these tests, I'd kind of prefer to remove that no-longer-needed test once we had complete coverage in the next diff anyway.

Yeah, I agree that would be a lot of movement for no real benefit. Let's just stick with the current tests in this PR then.

> I find this a bit confusing, so I think this might just be a stylistic difference. My main concern with this approach though would be having the list of tests hard-coded, since it leaves a lot of room for missing a case. 

I agree there are pros and cons with each case and I am fine with closing this thread with a "this is a difference in styles" conclusion